src/main/java/com/google/devtools/build/lib/rules/proto/ProtoInfo.java

```diff
@@ -0,0 +1,64 @@
+
+package com.google.devtools.build.lib.rules.proto;
+
+import com.google.devtools.build.lib.actions.Artifact;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import com.google.devtools.build.lib.events.Location;
+import com.google.devtools.build.lib.skyframe.serialization.autocodec.AutoCodec;
+import com.google.devtools.build.lib.skyframe.serialization.autocodec.AutoCodec.VisibleForSerialization;
+import com.google.devtools.build.lib.skylarkbuildapi.proto.ProtoInfoApi;
+import com.google.devtools.build.lib.syntax.Environment;
+import com.google.devtools.build.lib.syntax.EvalException;
+import com.google.devtools.build.lib.packages.BuiltinProvider;
+import com.google.devtools.build.lib.packages.NativeInfo;
+
+@Immutable
+@AutoCodec
+public final class ProtoInfo extends NativeInfo implements ProtoInfoApi<Artifact> {
```

Comment:
Why do you need to create a new provider? There is already one that tells which source files a `proto_library` provides, `ProtoSourcesProvider`.

Reply:
`ProtoSourcesProvider` doesn't seem to be exactly what I want here -- it describes only the .proto source files, but doesn't have information at the target level.

I'm thinking a `ProtoInfo` provider would have a `srcs :: ProtoSourcesProvider`, and also provide fields to discover what the prefix handling should be.

Alternatively, we could try adding info into `ProtoSourcesProvider` such that each source is no longer just a `file`, but a `(path, file)` tuple.

Reply:
Ooh, looking through this again I remember the other reasons I started on a new provider:

* `ProtoSourcesProvider` is an "old-style" provider, so it's not possible to return it from a Skylark rule (https://github.com/bazelbuild/bazel/issues/3701).
* The relationship between `ProtoSourcesProviderApi` and `ProtoSourcesProvider` is pretty unclear to me -- why is there this separation? The former is generic across implementations of `FileApi`, are there any cases where we'd use something other than `Artifact` as the file type of a protobuf library target?

I suspect the second is more of an Enterprise™ Java™ Best™ Practice™ thing vs a Bazel-specific issue, but it does aggravate the lack of documentation when I'm not able to figure out why there are multiple Java types that seem related to the same Skylark type.

Reply:
Well, I'm also somewhat bemused at the presence of `*Api` classes, but I assume the Starlark folks know what they are doing, so let's follow that example.

I think that the best course of action here is to turn `ProtoSourcesProvider` into a new-style provider. The old and new styles can coexist for an indefinite period of time so you would not need to do the shave the "deprecate the old-style API for `ProtoSourcesProvider`" yak. Adding the new-style (provider-based) API does require a bit of boilerplate, but following the footsteps of `JavaRuntimeInfo` and `JavaInfo` should be pretty straightforward: you make `ProtoSourcesProvider` a subclass of `BuiltinProvider`, add a static field called `PROVIDER` to it, then add a method to `proto_common` (just like `JavaSkylarkCommon#getJavaProvider`) that returns the provider.

(Unfortunately, neither `JavaInfo` nor `JavaRuntimeInfo` is 100% correct: the former makes `JavaInfo` a top-level symbol, which is wrong, and the latter uses one of the deprecated APIs for Skylark providers defined in Java)

Looking at the code,  there are plenty of places where we add the result of `ProtoSourcesProvider#getTransitiveProtoSources()` to nested sets so it's probably better to either another field returning a nested set of (path, artifact) tuples or, even better, that only for those artifacts whose path is different from the default one. Nested sets can contain anything that has a `#equals()` / `#hashCode()` method. In particular, `AutoValue` works.